service/worker/scheduler/workflow.go

```diff
@@ -867,7 +875,12 @@ func (s *scheduler) processBuffer() bool {
 		}
 	}
 
-	return tryAgain
+	if s.Schedule.State.RemainingActions == 0 && s.tweakables.Version >= FinishScheduleWithNoActionsLeft {
```

Comment:
I think the RemainingActions is optional, which means it is 0 by default. I think we probably should do this in canTakeScheduledAction() where we decrement RemainingActions, we should check after decrement, if it is 0 and if s.tweakables.Version >= FinishScheduleWithNoActionsLeft, then we set shouldFinish = true.

Reply:
Very good point, I might need to check `LimitedActions` here to make sure this is a schedule with limited actions. 

Initially I did exactly what [you are suggesting ](https://github.com/temporalio/temporal/pull/4399/commits/3f805380b4731bee2ef79293bc682a78d28f40be) but that part of code was never reached and the schedule kept running. I am currently looking into this.

Reply:
hmmm, you need to check with David. I'm not sure if the schedule spec wants to run only once on a give calendar spec is using the same RemainingActions mechanism. 
Another way to detect the schedule is done is to check if there is no next run wake up timer.

Reply:
I think checking the timer is better, as long as we check for a few other cases like paused and empty spec. But we also need the api change so let's leave this until we make a decision on that

Reply:
I think it's not possible to rely on time to finish the schedule early. The reason is that `nextWakeupTime` is never `0` for when we run out of actions. We do not decrement the counter when generating the next wake up time, which means `RemainingActions` never reaches `0` before calling `processBuffer`. 
The time is set to `0` with a manual check in `sleep` (see [here](https://github.com/temporalio/temporal/blob/master/service/worker/scheduler/workflow.go#L477-L479)), where results in the workflow suspended waiting for signals or updates. 

So IMO a manual check for running out of actions right after `processBuffer` should work.

Reply:
synched with Yimin this morning and since we want to also include schedules that do not have any more jobs to run (not only the ones that run out of remaining actions), I also included checking whether `nextWakeupTime` is zero.